File: backend/routes/subscriptions.js
// backend/routes/subscriptions.js

const express = require('express');
const supabase = require('../config/supabase');
const protect = require('../middleware/authMiddleware');
const router = express.Router();

// ─── GET ALL SUBSCRIPTIONS ───────────────────────────

router.get('/', protect, async (req, res) => {
  try {
    const { data: subscriptions, error } = await supabase
      .from('subscriptions')
      .select('*')
      .eq('user_id', req.user.userId)
      .order('total_spent', { ascending: false });

    if (error) throw error;

    // ── Enrich each subscription with receipt gap info ──
    const today = new Date();
    
    const enriched = subscriptions.map(sub => {
      const lastReceipt = sub.last_receipt_date
        ? new Date(sub.last_receipt_date)
        : null;
        
      const daysSinceLastReceipt = lastReceipt
        ? Math.floor((today - lastReceipt) / (1000 * 60 * 60 * 24))
        : null;

      // Determine receipt status label
      let receiptStatus = 'unknown';
      if (daysSinceLastReceipt !== null) {
        if (daysSinceLastReceipt <= 35) receiptStatus = 'recent';
        else if (daysSinceLastReceipt <= 90) receiptStatus = 'moderate';
        else receiptStatus = 'long_gap';
      }

      // Days until next renewal
      let daysUntilRenewal = null;
      if (sub.next_renewal_date) {
        const renewal = new Date(sub.next_renewal_date);
        daysUntilRenewal = Math.ceil((renewal - today) / (1000 * 60 * 60 * 24));
      }

      return {
        ...sub,
        daysSinceLastReceipt,
        receiptStatus,
        daysUntilRenewal
      };
    });

    res.json({ subscriptions: enriched });

  } catch (err) {
    console.error('Get subscriptions error:', err);
    res.status(500).json({ error: 'Failed to fetch subscriptions' });
  }
});

// ─── GET SINGLE SUBSCRIPTION ─────────────────────────

router.get('/:id', protect, async (req, res) => {
  try {
    const { data: subscription } = await supabase
      .from('subscriptions')
      .select('*')
      .eq('id', req.params.id)
      .eq('user_id', req.user.userId)
      .single();

    if (!subscription) {
      return res.status(404).json({ error: 'Subscription not found' });
    }

    // Get receipt history
    const { data: receipts } = await supabase
      .from('receipts')
      .select('*')
      .eq('subscription_id', req.params.id)
      .order('receipt_date', { ascending: false })
      .limit(24);  // Last 24 receipts = 2 years of monthly

    res.json({
      subscription,
      receipts: receipts || []
    });

  } catch (err) {
    res.status(500).json({ error: 'Failed to fetch subscription' });
  }
});

// ─── ADD MANUAL SUBSCRIPTION ─────────────────────────

// For subscriptions not found in Gmail
router.post('/manual', protect, async (req, res) => {
  try {
    const { serviceName, amount, billingDate, category } = req.body;

    if (!serviceName || !amount) {
      return res.status(400).json({
        error: 'Service name and amount are required'
      });
    }

    const { data: sub, error } = await supabase
      .from('subscriptions')
      .insert({
        user_id: req.user.userId,
        service_name: serviceName,
        amount: parseFloat(amount),
        currency: 'INR',
        category: category || 'Other',
        next_renewal_date: billingDate || null,
        total_receipts: 0,
        total_spent: 0,
        source_gmail: 'manual'  // Mark as manually added
      })
      .select()
      .single();

    if (error) throw error;

    res.status(201).json({ subscription: sub });

  } catch (err) {
    res.status(500).json({ error: 'Failed to add subscription' });
  }
});

// ─── UPDATE SUBSCRIPTION STATUS ──────────────────────

// User can mark subscription as:
// active = I am using this
// cancelled = I have cancelled this
// reviewing = I need to check this
router.patch('/:id', protect, async (req, res) => {
  try {
    const { userMarked } = req.body;

    const validStatuses = ['active', 'cancelled', 'reviewing'];
    if (!validStatuses.includes(userMarked)) {
      return res.status(400).json({ error: 'Invalid status' });
    }

    await supabase
      .from('subscriptions')
      .update({
        user_marked: userMarked,
        updated_at: new Date().toISOString()
      })
      .eq('id', req.params.id)
      .eq('user_id', req.user.userId);

    res.json({ message: 'Updated successfully' });

  } catch (err) {
    res.status(500).json({ error: 'Failed to update' });
  }
});

// ─── DELETE SUBSCRIPTION ─────────────────────────────

router.delete('/:id', protect, async (req, res) => {
  try {
    // Delete related records first due to foreign keys
    await supabase
      .from('alerts')
      .delete()
      .eq('subscription_id', req.params.id);

    await supabase
      .from('receipts')
      .delete()
      .eq('subscription_id', req.params.id);

    await supabase
      .from('subscriptions')
      .delete()
      .eq('id', req.params.id)
      .eq('user_id', req.user.userId);

    res.json({ message: 'Deleted successfully' });

  } catch (err) {
    res.status(500).json({ error: 'Failed to delete' });
  }
});

module.exports = router;

File: backend/routes/subscriptions.py
# backend/routes/subscriptions.py
# Subscriptions CRUD — exact port of routes/subscriptions.js

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from typing import Optional
from datetime import datetime
from math import floor, ceil

from config.database import supabase
from middleware.auth import get_current_user

router = APIRouter()


class ManualSubscription(BaseModel):
    serviceName: str
    amount: float
    billingDate: Optional[str] = None
    category: Optional[str] = "Other"


class StatusUpdate(BaseModel):
    userMarked: str


# ─── GET ALL SUBSCRIPTIONS ───────────────────────────

@router.get("/")
async def get_subscriptions(user: dict = Depends(get_current_user)):
    try:
        result = supabase.table("subscriptions") \
            .select("*") \
            .eq("user_id", user["userId"]) \
            .order("total_spent", desc=True) \
            .execute()

        subscriptions = result.data or []
        today = datetime.utcnow()

        # Enrich each subscription with receipt gap info
        enriched = []
        for sub in subscriptions:
            last_receipt = None
            days_since = None
            if sub.get("last_receipt_date"):
                last_receipt = datetime.fromisoformat(sub["last_receipt_date"])
                days_since = floor((today - last_receipt).total_seconds() / 86400)

            receipt_status = "unknown"
            if days_since is not None:
                if days_since <= 35:
                    receipt_status = "recent"
                elif days_since <= 90:
                    receipt_status = "moderate"
                else:
                    receipt_status = "long_gap"

            days_until_renewal = None
            if sub.get("next_renewal_date"):
                renewal = datetime.fromisoformat(sub["next_renewal_date"])
                days_until_renewal = ceil((renewal - today).total_seconds() / 86400)

            enriched.append({
                **sub,
                "daysSinceLastReceipt": days_since,
                "receiptStatus": receipt_status,
                "daysUntilRenewal": days_until_renewal
            })

        return {"subscriptions": enriched}

    except Exception as err:
        print(f"Get subscriptions error: {err}")
        raise HTTPException(status_code=500, detail="Failed to fetch subscriptions")


# ─── GET SINGLE SUBSCRIPTION ─────────────────────────

@router.get("/{sub_id}")
async def get_subscription(sub_id: str, user: dict = Depends(get_current_user)):
    try:
        sub_result = supabase.table("subscriptions") \
            .select("*") \
            .eq("id", sub_id) \
            .eq("user_id", user["userId"]) \
            .execute()

        if not sub_result.data or len(sub_result.data) == 0:
            raise HTTPException(status_code=404, detail="Subscription not found")

        receipts_result = supabase.table("receipts") \
            .select("*") \
            .eq("subscription_id", sub_id) \
            .order("receipt_date", desc=True) \
            .limit(24) \
            .execute()

        return {
            "subscription": sub_result.data[0],
            "receipts": receipts_result.data or []
        }

    except HTTPException:
        raise
    except Exception:
        raise HTTPException(status_code=500, detail="Failed to fetch subscription")


# ─── ADD MANUAL SUBSCRIPTION ─────────────────────────

@router.post("/manual")
async def add_manual(body: ManualSubscription, user: dict = Depends(get_current_user)):
    try:
        if not body.serviceName or not body.amount:
            raise HTTPException(status_code=400, detail="Service name and amount are required")

        result = supabase.table("subscriptions").insert({
            "user_id": user["userId"],
            "service_name": body.serviceName,
            "amount": float(body.amount),
            "currency": "INR",
            "category": body.category or "Other",
            "next_renewal_date": body.billingDate,
            "total_receipts": 0,
            "total_spent": 0,
            "source_gmail": "manual"
        }).select().execute()

        if not result.data:
            raise HTTPException(status_code=500, detail="Failed to add subscription")

        return {"subscription": result.data[0]}

    except HTTPException:
        raise
    except Exception:
        raise HTTPException(status_code=500, detail="Failed to add subscription")


# ─── UPDATE SUBSCRIPTION STATUS ──────────────────────

@router.patch("/{sub_id}")
async def update_status(sub_id: str, body: StatusUpdate, user: dict = Depends(get_current_user)):
    try:
        valid_statuses = ["active", "cancelled", "reviewing"]
        if body.userMarked not in valid_statuses:
            raise HTTPException(status_code=400, detail="Invalid status")

        supabase.table("subscriptions").update({
            "user_marked": body.userMarked,
            "updated_at": datetime.utcnow().isoformat()
        }).eq("id", sub_id).eq("user_id", user["userId"]).execute()

        return {"message": "Updated successfully"}

    except HTTPException:
        raise
    except Exception:
        raise HTTPException(status_code=500, detail="Failed to update")


# ─── DELETE SUBSCRIPTION ─────────────────────────────

@router.delete("/{sub_id}")
async def delete_subscription(sub_id: str, user: dict = Depends(get_current_user)):
    try:
        supabase.table("alerts").delete().eq("subscription_id", sub_id).execute()
        supabase.table("receipts").delete().eq("subscription_id", sub_id).execute()
        supabase.table("subscriptions").delete() \
            .eq("id", sub_id) \
            .eq("user_id", user["userId"]).execute()

        return {"message": "Deleted successfully"}

    except Exception:
        raise HTTPException(status_code=500, detail="Failed to delete")
